libxw/src/libxw/game.py

```diff
@@ -0,0 +1,85 @@
+from dataclasses import dataclass
+from enum import Enum
+from enum import auto
+from typing import Sequence
+from typing import Mapping
+
+from .puzzle import CellIndex
+from .puzzle import ClueDirection
+from .puzzle import Puzzle
+
+
+class CellCheckedness(Enum):
+    NOT_CHECKED = auto()
+    INCORRECT = auto()
+    CORRECT = auto()
+
+
+@dataclass
+class CellState:
+    guessed_chars: Sequence[str]
+    checkedness: CellCheckedness
+
+
+@dataclass
+class Game:
+    puzzle: Puzzle
+    cursor: CellIndex
+    direction: ClueDirection
+    cell_states: Mapping[CellIndex, CellState]
+
+    @classmethod
+    def new(cls, puzzle: Puzzle) -> "Game":
+        return cls(
+            puzzle,
+            cursor=(0, 0),
+            direction=ClueDirection.ACROSS,
+            cell_states={
+                index: CellState(
+                    guessed_chars=[],
+                    checkedness=CellCheckedness.NOT_CHECKED,
+                )
+                for index, cell in puzzle.cells.items()
+                if cell is not Puzzle.BLACK_SQUARE
+            },
+        )
+
+    def change_direction(self) -> "Game":
```

Comment:
All of these APIs return new objects rather than mutating the existing one.  It's not super clear to me if we need this, but in general it makes it a lot easier to do things like multithreading down the line (since without mutation there are no data races).